### ISSUE TITLE
libzmq error handling?

### ISSUE THREAD

**Issue by nilsh** [zeromq/clrzmq#18]
It would be good to have a way to handle Exceptions thrown by libzmq in c#. Currently even using a wrong SocketType at the client would unrecoverably crash my server app ( "Assertion failed: msg_->flags & ZMQ_MSG_MORE (rep.cpp:81)" ).
Is this possible, and if yes how?

**Comment by Guthur** [comment]
There is talk of 0MQ providing this sort of socket connection validation in the future, so that it could gracefully handle connecting a SUB to a REP socket for example.

Unfortunately clrzmq2 could not feasibly provide this sort of validation.

Cheers,
Michael

**Comment by nilsh** [comment]
Ok sounds good, but this is only one issue of many more i've discovered which will immediatly crash my server app which originally was designed not to stop... I know handling native c++ exceptions with c# is not possible, but are there any workarounds (maybe a protection layer)?

**Comment by Guthur** [comment]
There is a number of situations where zeromq will crash the server, which is unfortunate and should hopefully get resolved.

The one example I am thinking of is where 2 clients connect with the same Identity, this will crash the server and there is little, if anything clrzmq2 can do about it, it's an issue with the underlying library. (note. this may have been resolved in 0MQ but it was still an issue in 2.0.x iirc)

The problem also isn't really C++ exceptions; there are a number of asserts in libzmq, and it is not possible to catch these, as far as I am aware. The example you posted is one of these assert.

Cheers,
Michael

**Comment by nilsh** [comment]
It also crashes only if the remote server becomes unavailable while the program is running or when starting the program with at least one server unavailable (ping: "Destination Net Unreachable"). I have 2 Sources connected with REQ.

"
Destination Net Unreachable
errno == ECONNREFUSED || errno == ECONNRESET || errno == ETIMEDOUT || errno == EHOSTUNREACH (tcp_connecter.cpp:300)

Native stacktrace:

```
    mono() [0x494e01]
    /lib64/libpthread.so.0(+0xf2e0) [0x2b3a5007c2e0]
    /lib64/libc.so.6(gsignal+0x35) [0x2b3a502bc9e5]
    /lib64/libc.so.6(abort+0x186) [0x2b3a502bdee6]
    /usr/lib64/libzmq.so(+0x27508) [0x2aaaada38508]
    /usr/lib64/libzmq.so(+0x2c26f) [0x2aaaada3d26f]
    /usr/lib64/libzmq.so(+0x15d93) [0x2aaaada26d93]
    /usr/lib64/libzmq.so(+0x27d06) [0x2aaaada38d06]
    /lib64/libpthread.so.0(+0x6a4f) [0x2b3a50073a4f]
    /lib64/libc.so.6(clone+0x6d) [0x2b3a5035a82d]
```

Debug info from gdb:

# 

Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 

# used by your application.

Abort
"

is this behaviour expected? or should i post this on the general zeromq bugtracker? can i do anything to catch this error?
